README.md
# human_anatomy a Human Body part Selection package for Flutter

[![pub package](https://img.shields.io/pub/v/human_anatomy.svg)](https://pub.dartlang.org/packages/human_anatomy)

A Flutter package for iOS and Android for Selecting Human Body Part.
## Demo
<img src="http://sheikhsoft.com/screensort/human_anatomy.gif" width="340" height="640" title="Screen Shoot">


## Installation

First, add `human_anatomy` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

Second, add assets folder at the root your project

Then Add the flowing line in your pubspec.yaml like other image file

assets:
    - assets/head.svg
    - assets/left_ear.svg
    - assets/right_ear.svg
    - assets/left_eye.svg
    - assets/right_eye.svg
    - assets/nose.svg
    - assets/mouth.svg
    - assets/neck.svg
    - assets/chest.svg
    - assets/abdomin.svg
    - assets/pelvis.svg
    - assets/publs.svg
    - assets/left_soulder.svg
    - assets/right_soulder.svg
    - assets/left_arm.svg
    - assets/right_arm.svg
    - assets/left_elbow.svg
    - assets/right_elbow.svg
    - assets/left_forearm.svg
    - assets/right_forearm.svg
    - assets/left_wrist.svg
    - assets/right_wrist.svg
    - assets/left_hand.svg
    - assets/right_hand.svg
    - assets/left_thigh.svg
    - assets/right_thigh.svg
    - assets/left_knee.svg
    - assets/right_knee.svg
    - assets/left_leg.svg
    - assets/right_leg.svg
    - assets/left_ankle.svg
    - assets/right_ankle.svg
    - assets/left_foot.svg
    - assets/right_foot.svg

### iOS

No configuration required - the plugin should work out of the box.
### Android

No configuration required - the plugin should work out of the box.

### Code for the human_anatomy widget

``` dart
import 'package:human_anatomy/human_anatomy.dart';


class HumanAnatomyPage extends StatelessWidget {
  var _finalBodyPartList = [];
 
  @override
  Widget build(BuildContext context) {
    return HumanAnatomy(
    onChanged: bodyPartList,
    );
  }
  void bodyPartList(List value) {
      _finalBodyPartList = value;
    }
}
```

### Code for the Full page Human Anatomy widget

``` dart
import 'package:human_anatomy/full_page_human_anatomy.dart';

class FullPageHumanBody extends StatelessWidget {
  final Map mapData;
  const FullPageHumanBody({Key key, this.mapData}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return FullPageHumanAnatomy(      
      mapData: mapData,
      title: "Where it is?",
      route: "/fifth",
    );
  }
}
```


CHANGELOG.md
## [0.0.1]
* Initial release


lib/utils.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class Utils {
  showMyDialog(context, String text) {
    // flutter defined function

    showDialog(
        context: context,
        // return object of type AlertDialog
        child: new CupertinoAlertDialog(
          title: new Text("Massage"),
          content: new Text(text),
          actions: <Widget>[
            new FlatButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: new Text("OK"))
          ],
        ));
  }
}

class TopPrograssBar extends StatelessWidget {
  final String progressBarImagePath;

  TopPrograssBar({Key key, this.progressBarImagePath}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.only(top: 35.0),
      alignment: Alignment.topCenter,
      height: 25,
      child: Image.asset(progressBarImagePath, fit: BoxFit.fill),
    );
  }
}

class MyBackButton extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.bottomLeft,
      child: FlatButton.icon(
          onPressed: () {
            Navigator.of(context).pop();
          },
          icon: Icon(Icons.arrow_back),
          label: Text("Back")),
    );
  }
}

class TopTitle extends StatelessWidget {
  final double topMargin;
  final double leftMargin;
  final String title;

  const TopTitle({Key key, this.topMargin, this.leftMargin, this.title})
      : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.topLeft,
      margin: EdgeInsets.only(
        top: topMargin,
        left: leftMargin,
      ),
      child: Text(
        title,
        style: TextStyle(fontSize: 24.0, fontWeight: FontWeight.bold),
      ),
    );
  }
}

class SendButton extends StatelessWidget {
  final Function sendAction;
  final Icon buttonIcon;

  SendButton({Key key, this.sendAction, this.buttonIcon}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(10.0),
      alignment: Alignment.bottomRight,
      child: FloatingActionButton(
        backgroundColor: Colors.black87,
        child: Icon(Icons.arrow_forward),
        onPressed: () => sendAction,
      ),
    );
  }
}


lib/full_page_human_anatomy.dart
library human_anatomy;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:human_anatomy/utils.dart';

import 'human_anatomy.dart';

class FullPageHumanAnatomy extends StatelessWidget {
  var _finalBodyPartList = [];
  final Map mapData;
  final Widget container;
  final String title;
  final String route;

  FullPageHumanAnatomy(
      {Key key, this.mapData, this.container, this.title, this.route})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    print(mapData);
    return Scaffold(
      body: Container(
        child: Stack(
          children: <Widget>[
            container != null ? container : Container(),
            TopTitle(
              topMargin: 80.0,
              leftMargin: 50.0,
              title: title,
            ),
            Container(
              alignment: Alignment.center,
              margin: EdgeInsets.only(top: 50.0),
              child: HumanAnatomy(
                onChanged: bodyPartList,
              ),
            ),
            Container(
              padding: EdgeInsets.all(10.0),
              alignment: Alignment.bottomRight,
              child: FloatingActionButton(
                backgroundColor: Colors.black87,
                child: Icon(Icons.arrow_forward),
                onPressed: () {
                  mapData['bodyPart'] = _finalBodyPartList.toString();

                  _finalBodyPartList.length == 0
                      ? Utils().showMyDialog(
                          context, "Please Select Affected Body Part")
                      : Navigator.pushNamed(context, route, arguments: mapData);
                },
              ),
            ),
            MyBackButton(),
          ],
        ),
      ),
    );
  }

  void bodyPartList(List value) {
    _finalBodyPartList = value;
  }
}


lib/human_anatomy.dart
library human_anatomy;

import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

class HumanAnatomy extends StatefulWidget {
  final ValueChanged<List> onChanged;

  const HumanAnatomy({Key key, this.onChanged}) : super(key: key);
  @override
  _HumanAnatomyState createState() => new _HumanAnatomyState();
}

class _HumanAnatomyState extends State<HumanAnatomy> {
  var _bodyPartList = [];

  @override
  void initState() {
    super.initState();
    if (mounted) {
      _publishSelection(_bodyPartList);
    }
  }

  @override
  Widget build(BuildContext context) {
    return humanAnatomy();
  }

  Widget humanAnatomy() {
    return Container(
      width: 340.0,
      height: 557,
      child: SizedBox(
        child: Stack(
          children: <Widget>[
            bodyPart("assets/head.svg", "Head", 0.0, 0.0, 0.0, 70.0),
            bodyPart("assets/left_ear.svg", "leftEar", 32.0, 50.0, 0.0, 20.0),
            bodyPart("assets/right_ear.svg", "rightEar", 32.0, 0.0, 50.0, 20.0),
            bodyPart("assets/left_eye.svg", "LeftEye", 22.0, 20.0, 0.0, 10.0),
            bodyPart("assets/right_eye.svg", "rightEye", 22.0, 0.0, 20.0, 10.0),
            bodyPart("assets/nose.svg", "Nose", 25.0, 0.0, 0.0, 20.0),
            bodyPart("assets/mouth.svg", "Mouth", 48.0, 0.0, 0.0, 10.0),
            bodyPart("assets/neck.svg", "Neck", 58.0, 0.0, 0.0, 40.0),
            bodyPart("assets/chest.svg", "Chest", 93.0, 0.0, 0.0, 95.0),
            bodyPart("assets/abdomin.svg", "Abdomin", 175.0, 0.0, 0.0, 65.0),
            bodyPart("assets/pelvis.svg", "Pelvis", 225.0, 0.0, 0.0, 50.0),
            bodyPart("assets/publs.svg", "Publs", 275.0, 0.0, 0.0, 15.0),
            bodyPart("assets/left_soulder.svg", "leftSoulder", 90.0, 105.0, 0.0,
                50.0),
            bodyPart("assets/right_soulder.svg", "rightSoulder", 89.0, 0.0,
                110.0, 50.0),
            bodyPart("assets/left_arm.svg", "leftArm", 120.0, 145.0, 0.0, 70.0),
            bodyPart(
                "assets/right_arm.svg", "rightArm", 118.0, 0.0, 142.0, 70.0),
            bodyPart(
                "assets/left_elbow.svg", "leftElbow", 172.0, 172.0, 0.0, 39.0),
            bodyPart("assets/right_elbow.svg", "rightElbow", 170.0, 0.0, 170.0,
                40.0),
            bodyPart("assets/left_forearm.svg", "leftForearm", 195.0, 190.0,
                0.0, 54.0),
            bodyPart("assets/right_forearm.svg", "rightForearm", 195.0, 0.0,
                190.0, 54.0),
            bodyPart(
                "assets/left_wrist.svg", "leftWrist", 238.0, 220.0, 0.0, 23.0),
            bodyPart("assets/right_wrist.svg", "rightWrist", 238.0, 0.0, 220.0,
                23.0),
            bodyPart(
                "assets/left_hand.svg", "leftHand", 250.0, 250.0, 0.0, 60.0),
            bodyPart(
                "assets/right_hand.svg", "rightHand", 250.0, 0.0, 250.0, 60.0),
            bodyPart(
                "assets/left_thigh.svg", "leftThigh", 242.0, 63.0, 0.0, 138.0),
            bodyPart("assets/right_thigh.svg", "rightThigh", 242.0, 0.0, 63.0,
                138.0),
            bodyPart(
                "assets/left_knee.svg", "leftKnee", 360.0, 68.0, 0.0, 48.0),
            bodyPart(
                "assets/right_knee.svg", "rightKnee", 360.0, 0.0, 68.0, 48.0),
            bodyPart("assets/left_leg.svg", "leftLeg", 395.0, 64.0, 0.0, 105.0),
            bodyPart(
                "assets/right_leg.svg", "rightLeg", 393.0, 0.0, 65.0, 106.0),
            bodyPart(
                "assets/left_ankle.svg", "leftAnkle", 495.0, 64.0, 0.0, 25.0),
            bodyPart(
                "assets/right_ankle.svg", "rightAnkle", 493.0, 0.0, 68.0, 25.0),
            bodyPart(
                "assets/left_foot.svg", "leftFoot", 510.0, 80.0, 0.0, 60.0),
            bodyPart(
                "assets/right_foot.svg", "rightFoot", 508.0, 0.0, 83.0, 56.0),
          ],
        ),
      ),
    );
  }

  Widget bodyPart(String svgPath, String svgName, double marginTop,
      double marginRight, double marginLeft, double height) {
    Color _svgColor = _bodyPartList.contains(svgName) ? Colors.redAccent : null;
    final Widget bodyPartSvg = new SvgPicture.asset(svgPath,
        semanticsLabel: svgName, color: _svgColor);
    return Container(
      margin:
          EdgeInsets.only(top: marginTop, right: marginRight, left: marginLeft),
      height: height,
      alignment: Alignment.topCenter,
      child: GestureDetector(
          onTap: () {
            setState(() {
              if (_bodyPartList.contains(svgName)) {
                _bodyPartList.remove(svgName);
              } else {
                _bodyPartList.add(svgName);
                // _bodyPartList.forEach((e) => print(e));
              }
            });
          },
          child: bodyPartSvg),
    );
  }

  void _publishSelection(List _bodyPartList) {
    if (widget.onChanged != null) {
      widget.onChanged(_bodyPartList);
    }
  }
}
